selfdrive/locationd/locationd.cc

```diff
@@ -0,0 +1,378 @@
+    this->calib_from_device = this->device_from_calib.transpose();
+    this->calibrated = log.getCalStatus() == 1;
+  }
+}
+
+void Localizer::reset_kalman(double current_time) {
+  VectorXd init_x = this->kf->get_initial_x();
+  this->reset_kalman(current_time, init_x.segment<4>(3), init_x.head(3));
+}
+
+void Localizer::reset_kalman(double current_time, VectorXd init_orient, VectorXd init_pos) {
+  // too nonlinear to init on completely wrong
+  VectorXd init_x = this->kf->get_initial_x();
+  MatrixXdr init_P = this->kf->get_initial_P();
+  init_x.segment<4>(3) = init_orient;
+  init_x.head(3) = init_pos;
+
+  this->kf->init_state(init_x, init_P, current_time);
+
+  this->gyro_counter = 0;
+  this->acc_counter = 0;
+  this->speed_counter = 0;
+  this->cam_counter = 0;
+}
+
+void Localizer::handle_msg_bytes(const char *data, const size_t size) {
+  AlignedBuffer aligned_buf;
+
+  capnp::FlatArrayMessageReader cmsg(aligned_buf.align(data, size));
+  cereal::Event::Reader event = cmsg.getRoot<cereal::Event>();
+
+  this->handle_msg(event);
+}
+
+void Localizer::handle_msg(const cereal::Event::Reader& log) {
+  double t = log.getLogMonoTime() * 1e-9;
+  if (log.isSensorEvents()) {
+    this->handle_sensors(t, log.getSensorEvents());
+  } else if (log.isGpsLocationExternal()) {
+    this->handle_gps(t, log.getGpsLocationExternal());
+  } else if (log.isCarState()) {
+    this->handle_car_state(t, log.getCarState());
+  } else if (log.isCameraOdometry()) {
+    this->handle_cam_odo(t, log.getCameraOdometry());
+  } else if (log.isLiveCalibration()) {
+    this->handle_live_calib(t, log.getLiveCalibration());
+  }
+}
+
+kj::ArrayPtr<capnp::byte> Localizer::getMessageBytes(MessageBuilder& msg_builder, uint64_t logMonoTime,
+  bool inputsOK, bool sensorsOK, bool gpsOK)
+{
+  cereal::Event::Builder evt = msg_builder.initEvent();
+  evt.setLogMonoTime(logMonoTime);
+  cereal::LiveLocationKalman::Builder liveLoc = evt.initLiveLocationKalman();
+  this->buildLiveLocation(liveLoc);
+  liveLoc.setInputsOK(inputsOK);
+  liveLoc.setSensorsOK(sensorsOK);
+  liveLoc.setGpsOK(gpsOK);
+  return msg_builder.toBytes();
+}
+
+int Localizer::locationd_thread() {
+  const std::initializer_list<const char *> service_list =
+      { "gpsLocationExternal", "sensorEvents", "cameraOdometry", "liveCalibration", "carState" };
+  SubMaster sm(service_list, nullptr, { "gpsLocationExternal" });
+  PubMaster pm({ "liveLocationKalman" });
+
+  Params params;
+
+  while (!do_exit) {
+    sm.update();
+    for (const char* service : service_list) {
+      if (sm.updated(service) && sm.valid(service)) {
+        const cereal::Event::Reader log = sm[service];
+        this->handle_msg(log);
+      }
+    }
+
+    if (sm.updated("cameraOdometry")) {
+      uint64_t logMonoTime = sm["cameraOdometry"].getLogMonoTime();
+      bool inputsOK = sm.allAliveAndValid();
+      bool sensorsOK = sm.alive("sensorEvents") && sm.valid("sensorEvents");
```

Comment:
```suggestion
      bool sensorsOK = sm.allAliveAndValid("sensorEvents");
```

Reply:
after merging this, we can probably just deprecate the field unless we want to actually check the sensor values at some point